solr/solrj/src/java/org/apache/solr/client/solrj/impl/CloudSolrClient.java

```diff
@@ -1213,10 +1193,6 @@ protected DocCollection getDocCollection(String collection, Integer expectedVers
       // no such collection exists
       return null;
     }
-    if (!ref.isLazilyLoaded()) {
```

Comment:
Any chance this 
Ethos is now no longer used anywhere??

Reply:
I don't know what that means, but any way, I find this aspect of the PR/change most mysterious.  It appears this check is an optimization.  What effect does removing it have?

Reply:
> The CloudSolrClient only caches DocCollection instances where ref.isLazilyLoaded() returns true. **However, for CollectionRef instances handled by HTTP-based CSP, the default method CollectionRef.isLazilyLoaded() returns false.** This PR disables the check in CloudSolrClient.getDocCollection() to ensure caching works regardless of the ref.isLazilyLoaded return value.

(I added emphasis).  This is surprising.

Reply:
In summary, by bailing early here, we never even populated the cache.  The cache may not be important for a the ZK CSP because that has its own cache via ZkStateReader holding and always updating the ClusterState.  But this isn't true of HTTP CSP and I don't think layering another cache there is the ideal approach either.  The "stateVer" protocol in place makes sense; embraces an eventually-consistent mindset instead of pretending everything everywhere knows all.

By removing this check we'll get a _cached_ DocCollection, even for the ZK CSP which has access to the latest state (albeit eventually consistent).  This is kind of sad, removing a benefit of the ZK CSP in that it's routing is extremely closely up to date.

Perhaps instead, put this back and modify HttpClusterStateProvider to return a ClusterState.CollectionRef subclass that returns "true" for isLazilyLoaded, and that which actually fetches the state in get() (thus lazily not eagerly)?

Reply:
> In summary, by bailing early here, we never even populated the cache. The cache may not be important for a the ZK CSP because that has its own cache via ZkStateReader holding and always updating the ClusterState. But this isn't true of HTTP CSP and I don't think layering another cache there is the ideal approach either. The "stateVer" protocol in place makes sense; embraces an eventually-consistent mindset instead of pretending everything everywhere knows all.
That is correct.

> By removing this check we'll get a _cached_ DocCollection, even for the ZK CSP which has access to the latest state (albeit eventually consistent). This is kind of sad, removing a benefit of the ZK CSP in that it's routing is extremely closely up to date.

I believe ZK CSP always has its DocCollections cached in CSC since its works with _Lazy_ refs. However, the `collectionStateCache` in CSC has a `TTL` of 60 seconds, I believe that is when this cache is updated to have the "latest" state (get() on the lazy ref is invoked).

> Perhaps instead, put this back and modify HttpClusterStateProvider to return a ClusterState.CollectionRef subclass that returns "true" for isLazilyLoaded, and that which actually fetches the state in get() (thus lazily not eagerly)?

I tried this approach initially, but found that this cannot be done without adding another layer of cache in Http CSP.

Reply:
BTW, I uncovered that this lazy check was introduced in SOLR-6521 as part of a performance improvement relating to ensuring that concurrent cache lookups of an expired entry don't redundantly fetch it in parallel. 

> I believe ZK CSP always has its DocCollections cached in CSP since its works with Lazy refs

I confirmed this as well in a debugger.  It's because a ZkStateReader created *for* ZK CSP doesn't watch the state of any single collection (albeit it does watch the parent /collections path), whereas a ZkStateReader created in ZkController (i.e. on the server) does watch individual collections for local replicas.

Two questions upon us:  do we
* (A) remove !isLazilyLoaded short-circuit.  Basically all users use ZK CSP and this doesn't trigger for that.  We fixed the HTTP CSP performance interaction here so it can be removed.
* (B) keep the short-circuit, and make HTTP CSP return a CollectionRef that is lazy

I actually think BOTH :-)
Just (A) is tempting; it "works" and addresses the obvious perf regression.  However, note the comment "we MUST try to get a new version" and "this is a call to ZK" some lines below.  This code is written around the idea that the potentially expensive part is down there where we hold a lock (SOLR-6521).  Code archeology here was helpful to uncover the intent.  If we don't make it lazy, we miss the point and then SOLR-6521 accomplishes nothing for the HTTP CSP scenario.  We should go as far as asserting that the ref.isLazilyLoaded and document getCollectionRef that it _must_ return "lazy" references.

Thinking of that lock and SOLR-6521, we could disregard the lazy/non-lazy issue and move `ClusterState.CollectionRef ref = getCollectionRef(collection);` from before the lock to right down into the lock before calling `get()` on it.  Then it doesn't matter if it's lazy or not; the real lookup (talking to either ZK or HTTP Solr) will be in there, which is the intent of SOLR-6521.  I played with this locally and tests pass.  Lets do this?

Reply:
I agree - the lazy/non-lazy check becomes irrelevant if collectionRef lookup happens inside the lock.